### PR TITLE
DrivAerML: LR fine-tuning around champion (4e-4, 4.5e-4, 5.5e-4)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The champion uses lr=5e-4. Prior #2873 tested 4.5e-4/5.5e-4/6e-4 and all were worse — but that was at a different config (with gc, earlier architecture). The current champion has no gc, no WD, T_max=30, 4L/512d. Fine-tuning LR ±10-20% at this exact config might find a slightly better optimum. Also test lr=3e-4 (larger step down) to check if the LR-loss curve is monotonic or has a local dip below the champion.

## Current Baselines

| Dataset | Metric | Value |
|---------|--------|-------|
| TF | val_primary/surface_pressure_mae | 22.537 |
| TFP | val_primary/field_mse | 0.00434 |
| AF | val_primary/surface_mse | 0.000482 |
| DM | val_primary/surface_rel_l2_pct | 3.997% |
| DM external target | val_primary/surface_rel_l2_pct | 3.71% |

## Runs

Base command (all runs):
```bash
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm_lr_finetune
```

**Run 1** — lr=4e-4:
```bash
<base> --lr 4e-4
```

**Run 2** — lr=4.5e-4:
```bash
<base> --lr 4.5e-4
```

**Run 3** — lr=5.5e-4:
```bash
<base> --lr 5.5e-4
```

**Run 4** — lr=3e-4 (larger step down, monotonicity check):
```bash
<base> --lr 3e-4
```

## What to Report

For each run: best `val_primary/surface_rel_l2_pct`, epoch achieved, and W&B run ID. A clean LR vs. loss table is helpful to see the shape of the curve around the champion point.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT piccolo:

## Results

### LR Fine-Tuning Sweep Summary

All runs used the champion config: 4L/512d/8H, AdamW, T_max=30, no-EMA, Fourier, batch_size=1, 50k surface points, SENPAI_MAX_EPOCHS=9999, 360-min budget.

| LR | Best `val_primary/surface_rel_l2_pct` | Best Epoch | Total Epochs | Test `surface_rel_l2_pct` | W&B Run ID |
|----|---------------------------------------|------------|--------------|---------------------------|------------|
| 3e-4 | **5.745%** | 271 | 451 | 6.718% | `73iaz6a5` |
| 4e-4 | 13.851% | 38 | 445 | 15.509% | `vvo4tj1t` |
| 4.5e-4 | 15.278% | 93 | 437 | 16.648% | `3m5bs4do` |
| **5e-4 (champion)** | **3.997%** | **467** | **~467** | **—** | **`bht6h42t`** |
| 5.5e-4 | 12.706% | 42 | 457 | 13.676% | `qvrsywc4` |

### Commands Used

```bash
# All runs share this base:
CUDA_VISIBLE_DEVICES=<N> SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group dm_lr_finetune --agent piccolo \
  --lr <LR> --wandb-name "piccolo/dm-lr-<LR>"

# GPU 0: --lr 3e-4, GPU 1: --lr 4e-4, GPU 2: --lr 4.5e-4, GPU 3: --lr 5.5e-4
```

### What Happened

**None of the 4 LR variants beat the champion (3.997% at lr=5e-4).** All were substantially worse:

1. **lr=3e-4 was the best alternative at 5.745%** (44% worse than champion), finding its basin at epoch 271. This suggests a secondary stable basin at lower LR, but it's significantly shallower than the champion's.

2. **lr=4e-4, 4.5e-4, and 5.5e-4 were dramatically worse (12.7–15.3%)**, all finding their best checkpoints very early (epochs 38–93) then diverging badly. These runs spent most of their training time oscillating at 40–55%.

3. **The LR–loss curve is NOT monotonic.** It has a sharp optimum at lr=5e-4. Both stepping down (4.5e-4, 4e-4) and stepping up (5.5e-4) cause severe degradation, with the middle values (4e-4, 4.5e-4) being worse than the extremes. lr=3e-4 is the only variant that maintained any reasonable training stability.

4. **The champion LR (5e-4) appears to sit at a critical point** where the learning rate is just right for T_max=30 cosine cycling at 4L/512d. Small perturbations (±10–20%) destroy the optimizer's ability to find deep basins during cosine annealing cycles, suggesting a narrow "resonance" between LR and T_max.

### Suggested Follow-ups

- The sharp LR sensitivity suggests the champion config is already at a finely-tuned optimum. Further LR sweeps are unlikely to find improvement.
- **lr=4.8e-4 or lr=5.2e-4** (±4% from champion) could be tested for completeness, but the dramatic cliff at ±10% suggests diminishing returns.
- More promising levers may be T_max (already explored), depth/width, or regularization (gc, WD) rather than LR fine-tuning.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3047 because ADVISOR: Closing — lr=5e-4 confirmed as a sharp optimum.

**Results:**
| LR | Best val% | vs baseline |
|-----|----------|-------------|
| 3e-4 | 5.745% | +44% worse |
| 4e-4 | 13.851% | +247% worse |
| 4.5e-4 | 15.278% | +282% worse |
| 5.5e-4 | 12.706% | +218% worse |

**Conclusion:** Even ±10% LR change destroys training stability. The loss landscape has a narrow resonance at lr=5e-4 with T_max=30. Your analysis is exactly right — further LR sweeps are not productive. The LR axis is exhausted for DrivAerML at 4L/512d.

Reassigning you to a new DM experiment..
